src/convert/convertContext.ts

```diff
@@ -1,391 +0,0 @@
-    // When recomposing children that are non-decomposed, read and cache the parent XML to prevent
-    // reading the parent source file (referenced in all child SourceComponents) multiple times.
-    let parentXml: JsonMap | undefined;
-    if (parent.type.strategies?.transformer === TransformerStrategy.NonDecomposed && parent.xml) {
-      parentXml = await parent.parseXml();
-      this.parsedXmlCache.set(parent.xml, parentXml);
-    }
-
-    const parentXmlObj =
-      parent.type.strategies?.recomposition === RecompositionStrategy.StartEmpty
-        ? {}
-        : parentXml ?? (await parent.parseXml());
-
-    for (const child of children) {
-      if (!child.parent) {
-        throw messages.createError('noParent', [child.fullName, child.type.name]);
-      }
-      const groupName = child.type.xmlElementName ?? child.type.directoryName;
-      const { name: parentName } = child.parent.type;
-      const childSourceComponent = child as SourceComponent;
-
-      let xmlObj: JsonMap | undefined;
-      if (parentXml) {
-        // If the xml file for the child is in the cache, use it. Otherwise
-        // read and cache the xml file that contains this child and use it.
-        if (childSourceComponent.xml && !this.parsedXmlCache.has(childSourceComponent.xml)) {
-          // TODO: can we safely parallelize this?
-          // eslint-disable-next-line no-await-in-loop
-          this.parsedXmlCache.set(childSourceComponent.xml, await parent.parseXml(childSourceComponent.xml));
-        }
-        xmlObj = childSourceComponent.parseFromParentXml(
-          this.parsedXmlCache.get(
-            ensureString(childSourceComponent.xml, `Child component ${child.fullName} has no xml file`)
-          )
-        );
-      } else {
-        // TODO: can we safely parallelize this?
-        // eslint-disable-next-line no-await-in-loop
-        xmlObj = await childSourceComponent.parseXml();
-      }
-      if (!xmlObj) {
-        throw messages.createError('error_parsing_xml', [child.fullName, child.type.name]);
-      }
-      const childContents = xmlObj[child.type.name] ?? xmlObj;
-
-      if (!parentXmlObj[parentName]) {
-        parentXmlObj[parentName] = { [XML_NS_KEY]: XML_NS_URL };
-      }
-
-      // type safe way of checking childContents for the key
-      if (getString(childContents, XML_NS_KEY)) {
-        // child don't need to be written with `xmlns="http://soap.sforce.com/2006/04/metadata"` attribute
-        delete (childContents as JsonMap)[XML_NS_KEY];
-      }
-
-      const parentObj = parentXmlObj[parentName] as JsonMap;
-
-      if (!parentObj[groupName]) {
-        parentObj[groupName] = [];
-      }
-
-      // it might be an object and not an array.  Example: custom object with a Field property containing a single field
-      const group = ensureArray(parentObj[groupName]) as JsonArray;
-
-      group.push(childContents);
```

Comment:
mutation/side effect as the mechanism for setting values.  This took me a while to figure out why pushing to `group` would do anything.